compiler_gym/service/runtime/BenchmarkCache.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+#include "compiler_gym/service/runtime/BenchmarkCache.h"
+
+#include <glog/logging.h>
+
+using grpc::Status;
+using grpc::StatusCode;
+
+namespace compiler_gym::runtime {
+
+BenchmarkCache::BenchmarkCache(std::optional<std::mt19937_64> rand, size_t maxSizeInBytes)
+    : rand_(rand.has_value() ? *rand : std::mt19937_64(std::random_device()())),
+      maxSizeInBytes_(maxSizeInBytes),
+      sizeInBytes_(0){};
+
+const Benchmark* BenchmarkCache::get(const std::string& uri) const {
+  auto it = benchmarks_.find(uri);
+  if (it == benchmarks_.end()) {
+    return nullptr;
+  }
+
+  return &it->second;
+}
+
+void BenchmarkCache::add(const Benchmark&& benchmark) {
+  VLOG(3) << "Caching benchmark " << benchmark.uri() << ". Cache size = " << sizeInBytes()
+          << " bytes, " << size() << " items";
+
+  // Remove any existing value to keep the cache size consistent.
+  const auto it = benchmarks_.find(benchmark.uri());
+  if (it != benchmarks_.end()) {
+    const size_t replacedSize = it->second.ByteSizeLong();
+    benchmarks_.erase(it);
+    sizeInBytes_ -= replacedSize;
+  }
+
+  const size_t size = benchmark.ByteSizeLong();
+  if (sizeInBytes() + size > maxSizeInBytes()) {
+    if (size > maxSizeInBytes()) {
+      LOG(WARNING) << "Adding new benchmark with size " << size
+                   << " bytes exceeds total target cache size of " << maxSizeInBytes() << " bytes";
+    } else {
+      VLOG(3) << "Adding new benchmark with size " << size << " bytes exceeds maximum size "
+              << maxSizeInBytes() << " bytes, " << this->size() << " items";
+    }
+    prune();
+  }
+
+  benchmarks_.insert({benchmark.uri(), std::move(benchmark)});
+  sizeInBytes_ += size;
+}
+
+void BenchmarkCache::prune(std::optional<size_t> targetSize) {
+  int evicted = 0;
+  targetSize = targetSize.has_value() ? targetSize : maxSizeInBytes() / 2;
+
+  while (size() && sizeInBytes() > targetSize) {
+    // Select a benchmark randomly.
```

Comment:
Ignorable: I always thought the advantages of random over lru were that random doesn't need to keep the recently used list. If the data items are small that makes sense. Otherwise, isn't lru pretty much always better?

Reply:
I figured that there are two use common cases for this cache: (1) in a tight loop over a couple hundred benchmarks that will all fit in cache (2) over a massive set of training programs where the chance of a cache hit is negligible.

Given that, it didn't seem to me like LRU would be much of an advantage. I also found this interesting: "A random eviction policy degrades gracefully as the loop gets too big." from https://danluu.com/2choices-eviction/

Disclaimer: I know nothing about cache eviction policies and have utterly no idea what I'm talking about :)